src/lib.rs

```diff
@@ -0,0 +1,151 @@
+use k8s_openapi::api::core::v1::{self as apicore, Service};
+use k8s_openapi::Resource;
+use kubewarden_policy_sdk::host_capabilities::kubernetes::{
+    list_resources_by_namespace, ListResourcesByNamespaceRequest,
+};
+
+extern crate kubewarden_policy_sdk as kubewarden;
+use kubewarden::{protocol_version_guest, request::ValidationRequest, validate_settings};
+
+mod settings;
+use settings::Settings;
+
+use guest::prelude::*;
+use kubewarden_policy_sdk::wapc_guest as guest;
+
+#[no_mangle]
+pub extern "C" fn wapc_init() {
+    register_function("validate", validate);
+    register_function("validate_settings", validate_settings::<Settings>);
+    register_function("protocol_version", protocol_version_guest);
+}
+
+fn list_services(namespace: String) -> Result<Vec<Service>, String> {
+    let request = ListResourcesByNamespaceRequest {
+        api_version: "v1".to_owned(),
+        kind: Service::KIND.to_owned(),
+        namespace,
+        label_selector: None,
+        field_selector: None,
+    };
+    match list_resources_by_namespace(&request) {
+        Ok(response) => Ok(response.items),
+        Err(_) => Err("cannot list current services".to_owned()),
+    }
+}
+
+// Checks if the given services share the same set of selectors. Returns true
+// if both service the same selectors.
+fn is_service_with_duplicate_selector(service1: &Service, service2: &Service) -> bool {
+    if let Some(current_service_spec) = service1.spec.as_ref() {
+        if let Some(current_service_selector) = current_service_spec.selector.as_ref() {
+            if let Some(service_spec) = service2.spec.as_ref() {
+                if let Some(service_selector) = service_spec.selector.as_ref() {
+                    return current_service_selector == service_selector;
+                }
+            }
+        }
+    }
+    false
+}
+
+// Find the service that has the same selectors of the given service.
+// Returns the services name where the set of selectors match with the given service
+fn find_service_with_duplicate_selectors(
+    service: &Service,
+    current_services: Vec<Service>,
+) -> Vec<String> {
+    let mut duplicates = Vec::new();
+    for current_service in current_services.iter() {
+        if is_service_with_duplicate_selector(current_service, service) {
+            duplicates.push(current_service.metadata.name.clone().unwrap_or_default())
+        }
+    }
+    duplicates
+}
+
+fn validate_service(service: &Service) -> CallResult {
```

Comment:
I think this can be hugely simplified. The service `selector` is a `Option<BTreeMap<String, String>>`.

Both `Option` and `BTreeMap` implement the `Eq` traits. I think this can all be rewritten with `==` statements.
See [this gist](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e0e98b4cae33345c2761b93b43a49b02).

Keep in mind that a rejection happens only when the selector are truly equal. If a key has a different value, that's fine. If a map is a superset of another one, it's all fine

Reply:
Agree with the above comment